allennlp_rc/models/transformer_qa.py

```diff
@@ -0,0 +1,233 @@
+        )
+        span_end_logits = util.replace_masked_values(span_end_logits, possible_answer_mask, -1e32)
+        span_start_probs = torch.nn.functional.softmax(span_start_logits, dim=-1)
+        span_end_probs = torch.nn.functional.softmax(span_end_logits, dim=-1)
+        best_spans = get_best_span(span_start_logits, span_end_logits)
+        best_span_scores = torch.gather(
+            span_start_logits, 1, best_spans[:, 0].unsqueeze(1)
+        ) + torch.gather(span_end_logits, 1, best_spans[:, 1].unsqueeze(1))
+        best_span_scores = best_span_scores.squeeze(1)
+
+        output_dict = {
+            "span_start_logits": span_start_logits,
+            "span_start_probs": span_start_probs,
+            "span_end_logits": span_end_logits,
+            "span_end_probs": span_end_probs,
+            "best_span": best_spans,
+            "best_span_scores": best_span_scores,
+        }
+
+        # Compute the loss for training.
+        if answer_span is not None:
+            span_start = answer_span[:, 0]
+            span_end = answer_span[:, 1]
+            span_mask = span_start != -1
+            self._span_accuracy(
+                best_spans, answer_span, span_mask.unsqueeze(-1).expand_as(best_spans)
+            )
+
+            start_loss = cross_entropy(span_start_logits, span_start, ignore_index=-1)
+            if torch.any(start_loss > 1e9):
+                logger.critical("Start loss too high (%r)", start_loss)
+                logger.critical("span_start_logits: %r", span_start_logits)
+                logger.critical("span_start: %r", span_start)
+                assert False
+
+            end_loss = cross_entropy(span_end_logits, span_end, ignore_index=-1)
+            if torch.any(end_loss > 1e9):
+                logger.critical("Start loss too high (%r)", end_loss)
+                logger.critical("span_end_logits: %r", span_end_logits)
+                logger.critical("span_end: %r", span_end)
+                assert False
+
+            loss = (start_loss + end_loss) / 2
+
+            self._span_start_accuracy(span_start_logits, span_start, span_mask)
+            self._span_end_accuracy(span_end_logits, span_end, span_mask)
+
+            output_dict["loss"] = loss
+
+        # Compute the EM and F1 on SQuAD and add the tokenized input to the output.
+        if metadata is not None:
+            best_spans = best_spans.detach().cpu().numpy()
+
+            output_dict["best_span_str"] = []
+            context_tokens = []
+            for metadata_entry, best_span in zip(metadata, best_spans):
+                context_tokens_for_question = metadata_entry["context_tokens"]
+                context_tokens.append(context_tokens_for_question)
+
+                best_span -= 1 + len(metadata_entry["question_tokens"]) + 2
+                assert np.all(best_span >= 0)
+
+                predicted_start, predicted_end = tuple(best_span)
+                assert predicted_start >= 0
```

Comment:
Didn't you just have an assert that covers this three lines above?  And all of these asserts are just checking that `get_best_span` did its job correctly, which seems more appropriate to do as tests on `get_best_span` (which I believe we have), instead of having every caller of `get_best_span` inject its own asserts to make sure it's correct.

Reply:
The assert documents that it's safe to call `context_tokens_for_question[predicted_start]` below. The only way you would know this otherwise is by tracing the lineage of `predicted_span` through two other variables and into a function that gets called 50 lines earlier. It's not obvious.

Reply:
That is not how I read that assert, as its purpose is not exactly self-documenting.  Not sure why the reader is supposed to assume that the assert has anything to do with the call below. 
 A comment probably would be better documentation than a cryptic assert.  But, do what you think is best.

Reply:
I ended up removing the assert because upon careful reading, I realized the code is safe in any case. The assert checks for a thing that's re-checked by the `while` loop just below.